openhcl/virt_mshv_vtl/src/processor/snp/mod.rs

```diff
@@ -903,40 +901,46 @@ impl UhProcessor<'_, SnpBacked> {
     }
 
     async fn run_vp_snp(&mut self, dev: &impl CpuIo) -> Result<(), VpHaltReason<UhRunVpError>> {
-        // TODO CVM GUEST VSM: split out the "next vtl" that VTL 2 will exit to from
-        // last_vtl/current_vtl
-        let last_vtl = self
+        // TODO CVM GUEST VSM: actually check if there is an interrupt waiting
+        // for VTL 1 and switch to it if there is
+        let next_vtl = self
             .cvm_guest_vsm
             .as_ref()
-            .map_or(GuestVtl::Vtl0, |gvsm_state| gvsm_state.current_vtl);
+            .map_or(GuestVtl::Vtl0, |gvsm_state| gvsm_state.exit_vtl);
 
-        let mut vmsa = self.runner.vmsa_mut(last_vtl);
+        let mut vmsa = self.runner.vmsa_mut(next_vtl);
         let last_interrupt_ctrl = vmsa.v_intr_cntrl();
 
         vmsa.v_intr_cntrl_mut().set_guest_busy(false);
 
-        // TODO CVM GUEST VSM actually check and run vtl 1
         self.unlock_tlb_lock(Vtl::Vtl2);
-        let tlb_halt = self.should_halt_for_tlb_unlock(GuestVtl::Vtl0);
+        let tlb_halt = self.should_halt_for_tlb_unlock(next_vtl);
 
-        self.runner.set_halted(
-            self.backing.lapics[GuestVtl::Vtl0].halted
-                || self.backing.lapics[GuestVtl::Vtl0].startup_suspend
-                || tlb_halt,
-        );
+        let halt = self.backing.lapics[next_vtl].halted
+            || self.backing.lapics[next_vtl].startup_suspend // TODO GUEST VSM
```

Comment:
I left the todo there because I haven't looked enough into start vp and startup suspend yet to understand what it means for VTL 1--is it even allowed, what should happen, etc. So I wanted the TODO to make sure I go back and reason through it.

Reply:
Currently we'd expect it to never be set for VTL 1. VPs are created with startup_suspend = false, except for the BSP on VTL 0, and once VTL 1 is enabled all INITs are ignored. The latter rule may go away in the future, so this is purely future proofing.

Reply:
So to enable startup-suspend for VTL 1, we'd have to take explicit action to make startup_suspend = true on the non-BSP VPs or something? And presumably that would be bundled with the actual changes needed to make it work (is it handling the INITs for VTL 1 or something)?

Reply:
Yeah, we'd have to figure out what an INIT means in the context of multiple VTLs, since 'proper' handling would involve changing shared registers, but that might enable VTL 0 to attack VTL 1, etc etc